src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PreferIsExactSpellingIsTrueForKnownApisTest.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Testing;
+using Xunit;
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.InteropServices.PreferIsExactSpellingIsTrueForKnownApisAnalyzer,
+    Microsoft.NetCore.Analyzers.InteropServices.PreferIsExactSpellingIsTrueForKnownApisFixer>;
+namespace Microsoft.NetCore.Analyzers.InteropServices.UnitTests
+{
+    public sealed class PreferIsExactSpellingIsTrueForKnownApisTest
+    {
+        private DiagnosticResult CA1839_DefaultRule(int line, int column, params string[] arguments)
+           => VerifyCS.Diagnostic(PreferIsExactSpellingIsTrueForKnownApisAnalyzer.DefaultRule)
+               .WithLocation(line, column)
+               .WithArguments(arguments);
+
+        private DiagnosticResult CA1839_WideRule(int line, int column, params string[] arguments)
+           => VerifyCS.Diagnostic(PreferIsExactSpellingIsTrueForKnownApisAnalyzer.WideRule)
+               .WithLocation(line, column)
+               .WithArguments(arguments);
+
+        [Fact]
+        public async Task CA1839CSharpTest()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System.Runtime.InteropServices;
+
+public class C
+{
+    [DllImport(""user32.dll"")]
+    static extern void CallMsgFilter(); // should have exactSpelling and W rule
```

Comment:
Note that this going to call `CallMsgFilterA` method because the method is not marked as `CharSet = CharSet.Unicode`. Changing it to exactSpelling and W rule will change the behavior - whether this change change is ok or not depends on the rest of the arguments to the method.

The analyzer has to take `CharSet` into account as well.

Reply:
So the analyzer would only propse the fix in this case when the CharSet is set to unicode as well?

Reply:
> Would this not call CallMsgFilter? That method is also present.

`CallMsgFilter` is weird. It has all 3 variants: `CallMsgFilter`, `CallMsgFilterA`, `CallMsgFilterW`. `CallMsgFilter` is an alias for `CallMsgFilterA`. So this PInvoke as written ends up calling `CallMsgFilterA`.

It may be better to demonstrate the desired behavior on more regular methods that have just A and W variants.

Here are the searches that this is about: 
https://github.com/dotnet/runtime/blob/d1c7946bb0892005a58ec647e3a6d8f3514cdc6b/src/coreclr/src/vm/method.cpp#L5383
https://github.com/dotnet/runtime/blob/d1c7946bb0892005a58ec647e3a6d8f3514cdc6b/src/coreclr/src/vm/method.cpp#L5391

We care about situations when the first call to FindEntryPoint fails (wasting cycles) and the second call succeeds. The analyzer should be about fixing these situations. 

Suggesting fixes for the other situations would be a style suggestion. Consistent style may be still a fine idea, but it would not be help with performance.

Reply:
So if my ordinal would not end on A or W I would invoke A by default even if the ordinals don't match and these are the diagnostics we are interested in?

Reply:
> TestA and TestW. our method name is Test and no charset parameter is is present or it is CharSet.Ansi.

Yes.

> we have TestA and TestW. our method name is Test and Charset.Unicode parameter present.

This does not have a peformance problem today. The warning maybe nice-to-have for style, not for performance.

There is one additional case to warn for. It is the case in the original issue description:

we have TestW method. our method name is TestW and Charset.Unicode parameter present.

Reply:
I believe I am covering all cases now.

Reply:
I don’t believe this is resolved. Current analyzer implementation does not perform flow analysis, so cannot narrow down to cases that @jkotas mentioned where we only want to fire when first call fails and second succeeds. Current implementation is flow agnostic and would be a style suggestion, which does not seem to be the intent of this performance rule?

Reply:
We have all exported symbols in our `Dictionary<string,HashSet<string>>` . Suppose we have a method named `AbortSystemShutdown`, for which both `AbortSystemShutdownW` and `AbortSystemShutdownA` exist in our dictionary for `advapi32.dll`. As we have no attribute namedparameters present we default to `CharSet.Ansi`. As there is no known method for `AbortSystemShutdown` but for `AbortSystemShutdownA` the analyzer would warn as the first lookup fails. This is the targeted scenario is it not?

Reply:
This analyzer does not need to be flow agnostic.

>  we only want to fire when first call fails and second succeeds

The first and second call in my comment referred to calls to GetProcAddress method done internally in the runtime. The comment did not refer to calls done by user code that is being analyzed.

Reply:
Ah thanks for the clarification, sorry about confusion on my side.